### ISSUE TITLE
PVC still bound to spot instance node, creates issues with cluster-autoscaler

### ISSUE THREAD

**Issue by NissesSenap** [Azure/AKS#2991]
**What happened**:
Trying to automatically scale a spot instance, but is unable to do so thanks to a PVC being bound to an old node that doesn't exist any more.

**What you expected to happen**:

The PVC shouldn't be bound to the old node that doesn't exist in the cluster.

**How to reproduce it (as minimally and precisely as possible)**:

1. setup AKS cluster
2. setup a node pool with spot instances (don't think it needs to be spot).
3. Have a job running that triggers the spot instance creation and this job needs to be using a PVC. With affinity node selection that only allows usage of the spot instance.
4. When the job is done the autoscaler should remove the spot node since there is no more workload that it's needed to run.
5. The PVC is still bound to the spot instance
6. The job runs again and tries to spin up but it complains about the PVC still being bound to the old none existing node.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 1.22
- Size of cluster (how many worker nodes are in the cluster?): 12
- General description of workloads in the cluster (e.g. HTTP microservices, Java app, Ruby on Rails, machine learning, etc.): NA

**general Information**

Hi

This might be a cluster-autoscaler issue and I have created https://github.com/kubernetes/autoscaler/issues/4923 to look in to that as well.

We currently use thanos and together with it we use something called a compactor. The compactor runs using a cronjob and it got a cache disk which is a PVC.
To save money we use spot instances on AKS. If the job is not running the spot instances is shutdown.

The disk type that we are using is zrs which is multiple zone so the pvc shouldn't have any issues jumping around between the zones.

I'm getting an issue where my jobs can't start due to that the PVC is bound to a node that doesn't exist any more.

So when the cluster-autoscaler comes in and checks it tells me that a volume is in conflict.

`Normal   NotTriggerScaleUp  2m cluster-autoscaler  pod didn't trigger scale-up: 2 node(s) didn't match Pod's node affinity/selector, 1 node(s) had volume node affinity conflict`

I think this is thanks to the PVC shows as bound to a node that doesn't exist any more.
So as I see it, ether the volume should be unbound from the PVC or the cluster-autoscaler needs to ignore that the pod is bound to a existing node. But that might be a slippery slope.

You can view all the yaml and general information below. 

Below you can see that the volume is currently bound to an old memoryspot node.

k get pvc unbox-compactor-data2 -o yaml

```pvc.yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  annotations:
    meta.helm.sh/release-name: unbox
    meta.helm.sh/release-namespace: monitor
    pv.kubernetes.io/bind-completed: "yes"
    pv.kubernetes.io/bound-by-controller: "yes"
    volume.beta.kubernetes.io/storage-provisioner: disk.csi.azure.com
    volume.kubernetes.io/selected-node: aks-memoryspot1-11567145-vmss000003
  creationTimestamp: "2022-05-08T13:26:58Z"
  finalizers:
  - kubernetes.io/pvc-protection
  labels:
    app.kubernetes.io/instance: unbox
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: tenant
    app.kubernetes.io/part-of: unbox
    app.kubernetes.io/version: 1.16.0
    helm.sh/chart: tenant-0.3.1
    helm.toolkit.fluxcd.io/name: unbox
    helm.toolkit.fluxcd.io/namespace: monitor
  name: unbox-compactor-data2
  namespace: monitor
  resourceVersion: "373022994"
  uid: 5fe18e67-223d-462d-a43d-b96b8bdc6ffd
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 150Gi
  storageClassName: managed-csi-zrs
  volumeMode: Filesystem
  volumeName: pvc-5fe18e67-223d-462d-a43d-b96b8bdc6ffd
status:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 150Gi
  phase: Bound
```

As you can see below there is no spot instance node avaliable.

```shell
➜ k get nodes
NAME                                STATUS   ROLES   AGE   VERSION
aks-default-30194926-vmss000000     Ready    agent   24d   v1.21.9
aks-default-30194926-vmss000001     Ready    agent   24d   v1.21.9
aks-memory2-40909807-vmss000000     Ready    agent   22d   v1.21.9
aks-memory2-40909807-vmss000001     Ready    agent   22d   v1.21.9
aks-standard2-24926856-vmss000000   Ready    agent   24d   v1.21.9
aks-standard2-24926856-vmss000001   Ready    agent   24d   v1.21.9
aks-standard2-24926856-vmss000002   Ready    agent   24d   v1.21.9
aks-standard2-24926856-vmss000003   Ready    agent   24d   v1.21.9
aks-standard2-24926856-vmss000004   Ready    agent   24d   v1.21.9
aks-standard2-24926856-vmss000014   Ready    agent   22d   v1.21.9
aks-standard2-24926856-vmss000015   Ready    agent   22d   v1.21.9
aks-standard2-24926856-vmss000016   Ready    agent   22d   v1.21.9
aks-standard2-24926856-vmss000018   Ready    agent   22d   v1.21.9
aks-standard2-24926856-vmss000019   Ready    agent   22d   v1.21.9
```

The job that I'm using:

➜ k get job unbox-compactor-manuall -o yaml
```job.yaml
apiVersion: batch/v1
kind: Job
metadata:
  labels:
    aadpodidbinding: monitor
    app.kubernetes.io/component: compactor
    app.kubernetes.io/name: unbox
    app.kubernetes.io/part-of: unbox
    controller-uid: cc663554-1d32-430b-8975-d2cbb7395338
    job-name: unbox-compactor-manuall
  name: unbox-compactor-manuall
  namespace: monitor
spec:
  backoffLimit: 6
  completions: 1
  parallelism: 1
  selector:
    matchLabels:
      controller-uid: cc663554-1d32-430b-8975-d2cbb7395338
  template:
    metadata:
      creationTimestamp: null
      labels:
        aadpodidbinding: monitor
        app.kubernetes.io/component: compactor
        app.kubernetes.io/name: unbox
        app.kubernetes.io/part-of: unbox
        controller-uid: cc663554-1d32-430b-8975-d2cbb7395338
        job-name: unbox-compactor-manuall
    spec:
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: xkf.xenit.io/node-class
                operator: In
                values:
                - memory
            - matchExpressions:
              - key: kubernetes.azure.com/scalesetpriority
                operator: In
                values:
                - spot
      containers:
      - args:
        - compact
        - --data-dir=/tmp/data
        - --objstore.config-file=/etc/config/thanos.yaml
        - --retention.resolution-raw=10368000s
        - --retention.resolution-5m=10368000s
        - --retention.resolution-1h=10368000s
        - --compact.concurrency=1
        image: quay.io/thanos/thanos:v0.26.0
        imagePullPolicy: IfNotPresent
        name: compactor
        ports:
        - containerPort: 10902
          name: http
          protocol: TCP
        resources:
          limits:
            memory: 14Gi
          requests:
            cpu: "1"
            memory: 12Gi
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /etc/config/
          name: objectstore-secret
          readOnly: true
        - mountPath: /tmp/data
          name: data-volume
      dnsPolicy: ClusterFirst
      nodeSelector:
        kubernetes.azure.com/scalesetpriority: spot
        xkf.xenit.io/node-class: memory
      priorityClassName: tenant-low
      restartPolicy: OnFailure
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
      tolerations:
      - effect: NoSchedule
        key: kubernetes.azure.com/scalesetpriority
        operator: Equal
        value: spot
      volumes:
      - name: objectstore-secret
        secret:
          defaultMode: 420
          secretName: unbox-thanos-objstore-config
      - name: data-volume
        persistentVolumeClaim:
          claimName: unbox-compactor-data2
```

The whole describe output from the pod
```shell
➜ k describe pod  unbox-compactor-manuall-sn4m2
Name:                 unbox-compactor-manuall-sn4m2
Namespace:            monitor
Priority:             800000
Priority Class Name:  tenant-low
Node:                 <none>
Labels:               aadpodidbinding=monitor
                      app.kubernetes.io/component=compactor
                      app.kubernetes.io/name=unbox
                      app.kubernetes.io/part-of=unbox
                      controller-uid=cc663554-1d32-430b-8975-d2cbb7395338
                      job-name=unbox-compactor-manuall
Annotations:          <none>
Status:               Pending
IP:                  
IPs:                  <none>
Controlled By:        Job/unbox-compactor-manuall
Containers:
  compactor:
    Image:      quay.io/thanos/thanos:v0.26.0
    Port:       10902/TCP
    Host Port:  0/TCP
    Args:
      compact
      --data-dir=/tmp/data
      --objstore.config-file=/etc/config/thanos.yaml
      --retention.resolution-raw=10368000s
      --retention.resolution-5m=10368000s
      --retention.resolution-1h=10368000s
      --compact.concurrency=1
    Limits:
      memory:  14Gi
    Requests:
      cpu:        1
      memory:     12Gi
    Environment:  <none>
    Mounts:
      /etc/config/ from objectstore-secret (ro)
      /tmp/data from data-volume (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-kmxdk (ro)
Conditions:
  Type           Status
  PodScheduled   False
Volumes:
  objectstore-secret:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  unbox-thanos-objstore-config
    Optional:    false
  data-volume:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  unbox-compactor-data2
    ReadOnly:   false
  kube-api-access-kmxdk:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   Burstable
Node-Selectors:              kubernetes.azure.com/scalesetpriority=spot
                             xkf.xenit.io/node-class=memory
Tolerations:                 kubernetes.azure.com/scalesetpriority=spot:NoSchedule
                             node.kubernetes.io/memory-pressure:NoSchedule op=Exists
                             node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason             Age                     From                Message
  ----     ------             ----                    ----                -------
  Warning  FailedScheduling   7m52s                   default-scheduler   0/14 nodes are available: 12 node(s) didn't match Pod's node affinity/selector, 2 node(s) had taint {CriticalAddonsOnly: true}, that the pod didn't tolerate.
  Warning  FailedScheduling   6m40s (x1 over 7m51s)   default-scheduler   0/14 nodes are available: 12 node(s) didn't match Pod's node affinity/selector, 2 node(s) had taint {CriticalAddonsOnly: true}, that the pod didn't tolerate.
  Normal   NotTriggerScaleUp  2m40s (x20 over 7m42s)  cluster-autoscaler  pod didn't trigger scale-up: 2 node(s) didn't match Pod's node affinity/selector, 1 node(s) had volume node affinity conflict
```

**Comment by NissesSenap** [comment]
Good point, thanks @andyzhangx 

➜ k get pv pvc-5fe18e67-223d-462d-a43d-b96b8bdc6ffd -o yaml
```.yaml
apiVersion: v1
kind: PersistentVolume
metadata:
  annotations:
    pv.kubernetes.io/provisioned-by: disk.csi.azure.com
  creationTimestamp: "2022-05-08T13:27:25Z"
  finalizers:
  - kubernetes.io/pv-protection
  - external-attacher/disk-csi-azure-com
  name: pvc-5fe18e67-223d-462d-a43d-b96b8bdc6ffd
  resourceVersion: "345948185"
  uid: 61ba8c75-1815-4197-93a1-f54c051bc4eb
spec:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 150Gi
  claimRef:
    apiVersion: v1
    kind: PersistentVolumeClaim
    name: unbox-compactor-data2
    namespace: monitor
    resourceVersion: "345948101"
    uid: 5fe18e67-223d-462d-a43d-b96b8bdc6ffd
  csi:
    driver: disk.csi.azure.com
    volumeAttributes:
      csi.storage.k8s.io/pv/name: pvc-5fe18e67-223d-462d-a43d-b96b8bdc6ffd
      csi.storage.k8s.io/pvc/name: unbox-compactor-data2
      csi.storage.k8s.io/pvc/namespace: monitor
      requestedsizegib: "150"
      skuName: StandardSSD_ZRS
      storage.kubernetes.io/csiProvisionerIdentity: 1651944832699-8081-disk.csi.azure.com
    volumeHandle: /subscriptions/sub-id/resourceGroups/mc_rg-prod-we-aks_aks-prod-we-aks1_westeurope/providers/Microsoft.Compute/disks/pvc-5fe18e67-223d-462d-a43d-b96b8bdc6ffd
  nodeAffinity:
    required:
      nodeSelectorTerms:
      - matchExpressions:
        - key: topology.disk.csi.azure.com/zone
          operator: In
          values:
          - westeurope-1
      - matchExpressions:
        - key: topology.disk.csi.azure.com/zone
          operator: In
          values:
          - westeurope-2
      - matchExpressions:
        - key: topology.disk.csi.azure.com/zone
          operator: In
          values:
          - westeurope-3
      - matchExpressions:
        - key: topology.disk.csi.azure.com/zone
          operator: In
          values:
          - ""
  persistentVolumeReclaimPolicy: Delete
  storageClassName: managed-csi-zrs
  volumeMode: Filesystem
status:
  phase: Bound
```

**Comment by andyzhangx** [comment]
> `1 node(s) had volume node affinity conflict`

What's the node labels? "kubectl get no --show-labels"

**Comment by NissesSenap** [comment]
I removed a bunch of extra data. since it's the same nodes.
As you can see the spot node is currently not active in the cluster.

```shell
➜ kubectl get no --show-labels
NAME                                STATUS   ROLES   AGE   VERSION   LABELS
aks-default-30194926-vmss000000     Ready    agent   24d   v1.21.9   agentpool=default,beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=Standard_D2ds_v5,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=westeurope,failure-domain.beta.kubernetes.io/zone=westeurope-1,kubernetes.azure.com/agentpool=default,kubernetes.azure.com/cluster=MC_rg-prod-we-aks_aks-prod-we-aks1_westeurope,kubernetes.azure.com/mode=system,kubernetes.azure.com/node-image-version=AKSUbuntu-1804gen2containerd-2022.05.22,kubernetes.azure.com/os-sku=Ubuntu,kubernetes.azure.com/role=agent,kubernetes.io/arch=amd64,kubernetes.io/hostname=aks-default-30194926-vmss000000,kubernetes.io/os=linux,kubernetes.io/role=agent,node-role.kubernetes.io/agent=,node.kubernetes.io/instance-type=Standard_D2ds_v5,topology.disk.csi.azure.com/zone=westeurope-1,topology.kubernetes.io/region=westeurope,topology.kubernetes.io/zone=westeurope-1
aks-memory2-40909807-vmss000000     Ready    agent   22d   v1.21.9   agentpool=memory2,beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=Standard_E4ds_v5,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=westeurope,failure-domain.beta.kubernetes.io/zone=westeurope-2,kubernetes.azure.com/agentpool=memory2,kubernetes.azure.com/cluster=MC_rg-prod-we-aks_aks-prod-we-aks1_westeurope,kubernetes.azure.com/mode=user,kubernetes.azure.com/node-image-version=AKSUbuntu-1804gen2containerd-2022.05.22,kubernetes.azure.com/os-sku=Ubuntu,kubernetes.azure.com/role=agent,kubernetes.io/arch=amd64,kubernetes.io/hostname=aks-memory2-40909807-vmss000000,kubernetes.io/os=linux,kubernetes.io/role=agent,node-pool=memory2,node-role.kubernetes.io/agent=,node.kubernetes.io/instance-type=Standard_E4ds_v5,topology.disk.csi.azure.com/zone=westeurope-2,topology.kubernetes.io/region=westeurope,topology.kubernetes.io/zone=westeurope-2,xkf.xenit.io/node-class=memory
aks-standard2-24926856-vmss000019   Ready    agent   22d   v1.21.9   agentpool=standard2,beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=Standard_D4ds_v5,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=westeurope,failure-domain.beta.kubernetes.io/zone=westeurope-1,kubernetes.azure.com/agentpool=standard2,kubernetes.azure.com/cluster=MC_rg-prod-we-aks_aks-prod-we-aks1_westeurope,kubernetes.azure.com/mode=user,kubernetes.azure.com/node-image-version=AKSUbuntu-1804gen2containerd-2022.05.22,kubernetes.azure.com/os-sku=Ubuntu,kubernetes.azure.com/role=agent,kubernetes.io/arch=amd64,kubernetes.io/hostname=aks-standard2-24926856-vmss000019,kubernetes.io/os=linux,kubernetes.io/role=agent,node-pool=standard2,node-role.kubernetes.io/agent=,node.kubernetes.io/instance-type=Standard_D4ds_v5,topology.disk.csi.azure.com/zone=westeurope-1,topology.kubernetes.io/region=westeurope,topology.kubernetes.io/zone=westeurope-1
```

**Comment by andyzhangx** [comment]
so if you manually scale up the node count, it works? if so, then it's cluster-autoscaler issue.

**Comment by NissesSenap** [comment]
@andyzhangx replayed in my autoscaler issue as well: https://github.com/kubernetes/autoscaler/issues/4923 and pointed out that this hopefully will be solved soon if you are using k8s 1.22 in AKS.
Thus I will close this issue for now and to the upgrade and wait for a weeks and hopefully this will be solved then.
